Stop registering schemas for default objects

kajinamit opened the pull request:
We currently explicitly register schema of default objects in envtest but it seems this is not required. For example we do not register schema for rbacv1 api objects but it does not cause any problems.

Let's stop explicit registration. This allows us to avoid leaving the incomplete list and also helps us with maintaining the list more easily.

gibizer commented on controllers/placementapi_controller.go:
```diff
@@ -207,6 +207,19 @@ func (r *PlacementAPIReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		return ctrl.Result{}, err
+	}
+	if err == nil {
+		controllerutil.RemoveFinalizer(serviceDb, helper.GetFinalizer())
```
So lib-common adds the finalizer. Could you make it symmetric that lib-common does the removal too? I.e. we use the Database object for all the creation mechanism so lib-common hides how the MariaDBDatabase CR is handled. By doing "database.GetServiceDbByName" and getting back a MariaDBDatabase instance this abstraction becomes leaky. Can we move the finalizer removal also to the Database object to keep the abstraction?

pinikomarov replied:
@gibizer @stuggi how about this approach : 
- on the lib-common's side I've added the DeleteFinalizer func to the database module: 
https://github.com/pinikomarov/lib-common/blob/add_database_finalizers/modules/database/funcs.go#L262-L263
- Then in the placementapi_controller I've put a - database.DeleteFinalizer :
https://github.com/pinikomarov/placement-operator/blob/add_service_db_finalizer_deletion/controllers/placementapi_controller.go#L211

@stuggi I know we talked about returning a *Database obj for the DeleteFinalizer,
 but as the *Database struct has the *mariadbv1.MariaDBDatabase object in it:
https://github.com/pinikomarov/lib-common/blob/add_database_finalizers/modules/database/types.go#L29
```
type Database struct {
  database         *mariadbv1.MariaDBDatabase
  databaseHostname string
  databaseName     string
```
I'm hesitant to return it so we don't "leak" any unnecessary MariaDBDatabase instance as @gibizer put it, not sure about it though ..?

gibizer replied:
Returning a Database object that has a **private** mariadbv1.MariaDBDatabase field is totally OK as the client will not see / has no access to that private field. 

I'm not 100% what you would do with the returned Database instance in placement. If you plan to use it to call db.DeleteFinalizer() -on it, then be careful that such instance is created with NewDatabase but that is not part of  the code path of reconcileDelete(). Probably you can call NewDatabase in the reconcileDelete() code path too. So that is probably OK.

pinikomarov replied:
@stuggi , is there a reason to return a *Database in  https://github.com/pinikomarov/lib-common/blob/add_database_finalizers/modules/database/types.go#L29
?
we can just return an error (current commit) to handle any reconcile events ..

gibizer reviewed (changes requested):
Small and simple.

pinikomarov commented on controllers/placementapi_controller.go:
```diff
@@ -207,6 +207,12 @@ func (r *PlacementAPIReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	util.LogForObject(helper, "Reconciling Service delete", instance)
 
+	// remove db finalizer before the keystone one
+	db := database.GetDatabaseByName(ctx, helper, instance.Name)
```
ok

pinikomarov commented:
@stuggi @stuggi does this patch need a lgtm from me ? or a ok-to-test from you ?

SeanMooney commented on controllers/placementapi_controller.go:
```diff
@@ -207,6 +207,16 @@ func (r *PlacementAPIReconciler) SetupWithManager(mgr ctrl.Manager) error {
 func (r *PlacementAPIReconciler) reconcileDelete(ctx context.Context, instance *placementv1.PlacementAPI, helper *helper.Helper) (ctrl.Result, error) {
 	util.LogForObject(helper, "Reconciling Service delete", instance)
 
+	// remove db finalizer before the keystone one
```
im not really sure why this need to happen before the keystone one

gibizer replied:
I don't know why we would need to do this first. I only see a single rule: We need to remove all our finalizers from other CRs first *then at last* remove the finalizer from ourselves (PlacementAPI).

SeanMooney replied:
ya thats what i was thinking too so i was wondering why this comment as added as it implies the ordering is important between the db and keystone which i don't think it is.

just that we remove the finaliser on other resources before ourselves as you said.

pinikomarov replied:
right thanks , that was a typo , it should read : 
// remove db finalizer before the placement one
I updated the comment in the file

gibizer reviewed (dismissed):
looks good to me

pinikomarov commented:
@gibizer are we good here ? any issues ?

stuggi commented:
you have to update the co.mod to use the latest lib-common version with your change.

gibizer commented:
> @gibizer are we good here ? any issues ?

it looks good to me.

pinikomarov commented:
@gibizer not ready yet , I still have to update the go.mod thing - https://github.com/openstack-k8s-operators/placement-operator/pull/50#issuecomment-1319615936

pinikomarov commented:
@gibizer I _think_ it's ready now , add a commit with : 
`go get -u github.com/openstack-k8s-operators/lib-common/modules/database`

SeanMooney commented:
ignoring the fact that htis fails the checks you also have multiple commits with the same name and you create additional commits to address the feedback.

for example you have "use GetDatabaseByName" followed by "fix for GetDatabaseByName"

we do not use squash merges in this repo we try to use rebase merge where we can so every commit in the history should be independently correct.

can you do an interactive rebase and squash the fixup commit into the correct commits and make sure that the set of pathces are correct individually not just when the whole pr is merged as a unit.

in addition to that you need to address the failing tests
currently the call to delete the finalise is causing a segmentation fault

https://github.com/openstack-k8s-operators/placement-operator/actions/runs/3524854901/jobs/5921094815#step:8:24

my guess is that the finaliser has not been created in the test env so you need to update the tests to simulate that.

SeanMooney reviewed (changes requested):
see comment https://github.com/openstack-k8s-operators/placement-operator/pull/50#issuecomment-1324770801

there are several things wrong with this patch

pinikomarov commented:
tnx @SeanMooney I'll start on some fixing up

pinikomarov commented:
@gibizer @SeanMooney  I've simplified the commits , nothing to big here so a single commit fits this pr.
Can you please give the ok to test ? 

@gibizer about the failed tests : 

https://github.com/openstack-k8s-operators/placement-operator/actions/runs/3524854901/jobs/5921094815#step:8:24
i.e `FAIL	github.com/openstack-k8s-operators/placement-operator/controllers	6.068s`
```java
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1570e00]

goroutine 832 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:118 +0x1f4
panic({0x18252e0, 0x29d4ab0})
	/opt/hostedtoolcache/go/1.18.7/x64/src/runtime/panic.go:838 +0x207
github.com/openstack-k8s-operators/mariadb-operator/api/v1beta1.(*MariaDBDatabase).GetFinalizers(0x0?)
```
Is this a result a non existing finalizer that the test env tries to "get" ?
(because maybe since the git-workflows test env does not create it for the test like in the real env..)
So perhaps for that we should add a conditional for GetFinalizer only if it exists ..? 
like in the Keystone operator:
```go 
if !controllerutil.ContainsFinalizer(instance, helper.GetFinalizer())
```
https://github.com/pinikomarov/keystone-operator/blob/set_db_finalizer/controllers/keystoneapi_controller.go#L450-L451

Or is it something else ?

gibizer commented on controllers/placementapi_controller.go:
```diff
@@ -207,6 +207,16 @@ func (r *PlacementAPIReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	util.LogForObject(helper, "Reconciling Service delete", instance)
 
+	// remove db finalizer before the placement one
+	db, err := database.GetDatabaseByName(ctx, helper, instance.Name)
+	if err != nil && !k8s_errors.IsNotFound(err) {
+		return ctrl.Result{}, err
```
So this error handling ignores the case when MariaDBDatabase is not found,`!k8s_errors.IsNotFound(err)`. Therefore allows to db.database to be nil. This causes the panic when `db.DeleteFinalizer` is called. If MariaDBDatabase is not found it means placement controller haven't created it yet. So there is no finalizer to remove from it. You need to skip db.DeleteFinalizer call but do the rest of the finalizer removal.

```
	if !k8s_errors.IsNotFound(err) {
		if err := db.DeleteFinalizer(ctx, helper); err != nil {
			return ctrl.Result{}, err
		}
	}
```

alternatively you can change db.DeleteFinalize() in lib-common to handle the case when `db.database` is nil.

pinikomarov replied:
I think I'm ok with : 
	`if !k8s_errors.IsNotFound(err) {`  in controllers/placementapi_controller.go:216
Since error handling logic is currently on that side as it is now.		

also , if I would like to test gates locally , using openstack-k8s-operators-ci and the :ssh to repo dir and running "go test -mod=mod -v ./..." errors out with 

` msg: "failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory",`
Is there a way set up that local gate to run ?  against a running crc cluster perhaps ?

gibizer replied:
> also , if I would like to test gates locally , using openstack-k8s-operators-ci and the :ssh to repo dir and running "go test -mod=mod -v ./..." errors out with
> 
> ` msg: "failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory",` Is there a way set up that local gate to run ? against a running crc cluster perhaps ?

use `make test` as that make target also calls `make envtest` that does the env setup

SeanMooney replied:
this looks like it has been addressed

the new code will return if we get any error other then not found.

pinikomarov commented:
Any more issues with this patch ?

gibizer reviewed (approved):
Looks good to me.

gibizer commented:
> Any more issues with this patch ? I believe the env-test should pass now :
> 
> ```shell
> [r@r placement-operator_pkomarov]$ make test
> [..]
> 
> ok      github.com/openstack-k8s-operators/placement-operator/controllers       9.465s  coverage: 43.2% of statements
> ```
> 
> can you run workflows with oktotest please ?

Could you give me some pointers to oktotest? Why we need it? How it helps?

SeanMooney commented:
i belive its only requried for someone that has not commit to the repo before.

so when your first patch is merged i don't think you will be asked on placement again
you will be considered a contributor at that point. also /ok-to-test i belive is only needed for prow we are still using github actions for ci at the moment.
i don't think i have currently configured it to require approval to run test for new continutors.

kajinamit committed:
Stop registering schemas for default objects

We currently explicitly register schema of default objects in envtest
but it seems this is not required. For example we do not register
schema for rbacv1 api objects but it does not cause any problems.

Let's stop explicit registration. This allows us to avoid leaving
the incomplete list and also helps us with maintaining the list more
easily.

abays reviewed (approved):
/lgtm
